### PR TITLE
tests: transactional email sending and saving

### PR DESCRIPTION
**Pull request by tan-zx**
tests for transactional email sending and saving
- include tests as promised in https://github.com/opengovsg/postmangovsg/pull/1768
- include tests for API transactional emails to cover edge cases that were previously omitted (e.g. malicious attachments etc.)

includes some light refactoring

I originally created these files too, but as discussed with @stanleynguyen, these integration tests should be able to cover everything we need at this point

image

### CODE REVIEW

File: backend/src/email/routes/tests/email-transactional.routes.test.ts
Comment (tan-zx) [inline backend/src/email/routes/tests/email-transactional.routes.test.ts:34]: not very sure why this was chosen, but I had to change to `restoreAllMocks` to get rid of side effects so that tests running in series would pass. (before that, they would pass individually but not as a group.)

File: backend/src/email/routes/tests/email-transactional.routes.test.ts
Comment (tan-zx) [inline backend/src/email/routes/tests/email-transactional.routes.test.ts:159]: to remove before merging in (genuinely not sure how it's done and whether there's a lesson here for how I should write my tests differently if so)

File: backend/src/email/routes/tests/email-transactional.routes.test.ts
Comment (tan-zx) [inline backend/src/email/routes/tests/email-transactional.routes.test.ts:492]: oops to remove

File: backend/src/email/services/email-transactional.service.ts
Comment (tan-zx) [inline backend/src/email/services/email-transactional.service.ts:57]: refactored this for greater precision

File: backend/src/email/utils/query.ts
Comment (tan-zx) [inline backend/src/email/utils/query.ts:1]: deleted this and included in `EmailService`
- fits better there
- allows me to mock the return value

File: backend/src/email/routes/tests/email-transactional.routes.test.ts
Comment (tan-zx) [inline backend/src/email/routes/tests/email-transactional.routes.test.ts:477]: oops missed this, will do later

File: backend/src/core/services/file-attachment.service.ts
Comment (stanleynguyen) [inline backend/src/core/services/file-attachment.service.ts:59]: This function is in this file, we don't have to `FileAttachmentService.`

Comment (stanleynguyen) [reply backend/src/core/services/file-attachment.service.ts:59]: This is creating unnecessary depending on the `FileAttachmentService` object

File: backend/src/email/middlewares/email-transactional.middleware.ts
Comment (stanleynguyen) [inline backend/src/email/middlewares/email-transactional.middleware.ts:31]: Why are we changing this function name? Adding `Md5` and `From` doesn't provide any extra context. Because for the reader, I don't need to care about the underlying algo and preposition

File: backend/src/core/errors/transactional.errors.ts
Comment (stanleynguyen) [inline backend/src/core/errors/transactional.errors.ts:12]: This name change is misleading cos the original message is not `empty` it's `invalid`

File: backend/src/email/services/email-transactional.service.ts
Comment (stanleynguyen) [reply backend/src/email/services/email-transactional.service.ts:57]: See [above](https://github.com/opengovsg/postmangovsg/pull/1846/files#r1005439545)

File: backend/src/email/middlewares/email-transactional.middleware.ts
Comment (stanleynguyen) [reply backend/src/email/middlewares/email-transactional.middleware.ts:31]: Also, please don't export this function, u can just mock the underlying `crypto` package. This function is intentionally kept private

File: backend/src/email/routes/tests/email-transactional.routes.test.ts
Comment (stanleynguyen) [inline backend/src/email/routes/tests/email-transactional.routes.test.ts:42]: Please don't move this here. It's over-generalisation. For tests, the default approach should be sharing as little as possible. Test cases should be easily readable and understandable in isolation

Comment (stanleynguyen) [reply backend/src/email/routes/tests/email-transactional.routes.test.ts:42]: You need a very very very very good reason for generalisation in test cases

File: backend/src/email/routes/tests/email-transactional.routes.test.ts
Comment (stanleynguyen) [reply backend/src/email/routes/tests/email-transactional.routes.test.ts:159]: Because we're mocking the email sending, ofc it passes

File: backend/src/email/routes/tests/email-transactional.routes.test.ts
Comment (stanleynguyen) [inline backend/src/email/routes/tests/email-transactional.routes.test.ts:204]: This test case is testing way more than this title suggests, it's testing for the email being persisted in the db with the corresponding error as well

File: backend/src/email/routes/tests/email-transactional.routes.test.ts
Comment (stanleynguyen) [inline backend/src/email/routes/tests/email-transactional.routes.test.ts:284]: Can we be more specific about the `error` here?

File: backend/src/email/routes/tests/email-transactional.routes.test.ts
Comment (stanleynguyen) [inline backend/src/email/routes/tests/email-transactional.routes.test.ts:324]: Can we be more specific about the `error` here?

**Review by stanleynguyen — CHANGES_REQUESTED** [review]
image

File: backend/src/core/errors/transactional.errors.ts
Comment (tan-zx) [reply backend/src/core/errors/transactional.errors.ts:12]: so this error is only thrown if the message becomes empty after removing the invalid HTML tags

normally, if there are invalid HTML tags, they will just be automatically removed and this error won't be thrown. that's why I renamed it

(you can see that's what the original message was getting at `message = 'Message template is invalid as it only contains invalid HTML tags'` → it becomes empty after the invalid HTML tags are removed)

File: backend/src/core/services/file-attachment.service.ts
Comment (tan-zx) [reply backend/src/core/services/file-attachment.service.ts:59]: Ohh so the reason I wrapped it in the `FileAttachmentService` object is so that I can mock the value here:
image

Is there some other way I can achieve this without creating this dependency 🤔

File: backend/src/email/middlewares/email-transactional.middleware.ts
Comment (tan-zx) [reply backend/src/email/middlewares/email-transactional.middleware.ts:31]: > Also, please don't export this function, u can just mock the underlying crypto package. This function is intentionally kept private

Ohh that's a very good point, let me change it back.

Can you show me how to mock the underlying `crypto` package? I think I have seen you do it in a previous PR before, but I don't fully understand how that works.

File: backend/src/email/routes/tests/email-transactional.routes.test.ts
Comment (tan-zx) [reply backend/src/email/routes/tests/email-transactional.routes.test.ts:42]: Fair enough! I was trying to generalise the code because I saw a lot of repetition, but I can totally see this particular `mockSendEmail` as going one step too far.

File: backend/src/email/routes/tests/email-transactional.routes.test.ts
Comment (tan-zx) [reply backend/src/email/routes/tests/email-transactional.routes.test.ts:204]: OK will fix the title

File: backend/src/email/routes/tests/email-transactional.routes.test.ts
Comment (tan-zx) [reply backend/src/email/routes/tests/email-transactional.routes.test.ts:159]: Ohh I was confused why this tests passes and why the next test fails.

```TS
test('Should thrown an error with invalid custom from address', async () => {
    const res = await request(app)
      .post(endpoint)
      .set('Authorization', `Bearer ${apiKey}`)
      .send({
        ...validApiCall,
        from: 'Hello <invalid@agency.gov.sg>',
        reply_to: user.email,
      })

    expect(res.status).toBe(400)
    expect(mockSendEmail).not.toBeCalled()
})
```

But looking back I think I figured out what's going on.

I was expecting a db call to `EmailFromAddress` to check whether the the custom from address is in our database. This is used e.g. in `EmailMiddleware.existsFromAddress`, which is called for email campaigns.

In `email-transactional.routes.ts`, we call `emailMiddleware.isFromAddressAccepted`, which simply checks whether `(fromAddress !== userEmail && fromAddress !== defaultFromAddress)`. 

Per our previous discussion, we said it's enough to have an API for users to check that their custom from email address is acceptable and just throw them a 500 error if they pass the check above but fail at the AWS SES (see this [issue](https://github.com/opengovsg/postmangovsg/issues/1837)).

File: backend/src/email/routes/tests/email-transactional.routes.test.ts
Comment (tan-zx) [reply backend/src/email/routes/tests/email-transactional.routes.test.ts:284]: just checking, this and all the ones below are the same as [this comment](https://github.com/opengovsg/postmangovsg/pull/1846#discussion_r1005448762) right, which is to use a longer title to describe the test in full?

Comment (stanleynguyen) [reply backend/src/email/routes/tests/email-transactional.routes.test.ts:284]: yesser

File: backend/src/core/errors/transactional.errors.ts
Comment (stanleynguyen) [reply backend/src/core/errors/transactional.errors.ts:12]: > so this error is only thrown if the message becomes empty after removing the invalid HTML tags
> 
> normally, if there are invalid HTML tags, they will just be automatically removed and this error won't be thrown. that's why I renamed it
> 
> (you can see that's what the original message was getting at `message = 'Message template is invalid as it only contains invalid HTML tags'` → it becomes empty after the invalid HTML tags are removed)
> 
> to be exact, it could be either the subject or the body that is empty, but if at least either one is empty, the error will be thrown

Hmm but I feel that we don't need that kinda granularity here. `Invalid` would be more future-proof in case we add more rules to this validation

File: backend/src/email/middlewares/email-transactional.middleware.ts
Comment (stanleynguyen) [reply backend/src/email/middlewares/email-transactional.middleware.ts:31]: Use [this](https://jestjs.io/docs/manual-mocks)

File: backend/src/core/services/file-attachment.service.ts
Comment (stanleynguyen) [reply backend/src/core/services/file-attachment.service.ts:59]: > Ohh so the reason I wrapped it in the `FileAttachmentService` object is so that I can mock the value here: image
> 
> Is there some other way I can achieve this without creating this dependency 🤔

No u shouldn't be exporting a private function just so that u can mock it - this is a really really bad reason to export, mock the external dependency that it relies on

File: backend/src/email/routes/tests/email-transactional.routes.test.ts
Comment (tan-zx) [reply backend/src/email/routes/tests/email-transactional.routes.test.ts:42]: Oh I guess it's not in the sense that some of the test case codepath would've terminated before hitting the `sendEmail` part? But otherwise the tests all assume that `sendEmail` would be successful?

File: backend/src/core/errors/transactional.errors.ts
Comment (tan-zx) [reply backend/src/core/errors/transactional.errors.ts:12]: Hmm but I must say that in the beginning, when I was trying to cause this error before I looked at the code in detail, I assumed that the error would be thrown if invalid HTML tags were used or if the entire message was wrapped around invalid HTML tags.

Then after poking further, I realise that our code removed the HTML tags directly and only throws the error in the edge case that this removal causes the field to be empty.

I would prefer to be a bit more specific, but it's no big deal either way. Will defer to your decision.

File: backend/src/email/routes/tests/email-transactional.routes.test.ts
Comment (stanleynguyen) [reply backend/src/email/routes/tests/email-transactional.routes.test.ts:42]: > Oh I guess it's not in the sense that some of the test case codepath would've terminated before hitting the `sendEmail` part? But otherwise the tests all assume that `sendEmail` would be successful?

That's assumption is not true (or at least not for long). We dont have it here yet but we might need to handle errors from `sendEmail` and write tests for that

File: backend/src/core/errors/transactional.errors.ts
Comment (stanleynguyen) [reply backend/src/core/errors/transactional.errors.ts:12]: > Hmm but I must say that in the beginning, when I was trying to cause this error before I looked at the code in detail, I assumed that the error would be thrown if invalid HTML tags were used or if the entire message was wrapped around invalid HTML tags.
> 
> Then after poking further, I realise that our code removed the HTML tags directly and only throws the error in the edge case that this removal causes the field to be empty.
> 
> I would prefer to be a bit more specific, but it's no big deal either way. Will defer to your decision.

I think that `Empty` is actually more misleading, as the original message is **not** empty

File: backend/src/email/routes/tests/email-transactional.routes.test.ts
Comment (tan-zx) [reply backend/src/email/routes/tests/email-transactional.routes.test.ts:159]: Do you think this is something we should add? (Could be a separate PR, would handle the issue Kee Wei flagged as well)

File: backend/src/core/errors/transactional.errors.ts
Comment (tan-zx) [reply backend/src/core/errors/transactional.errors.ts:12]: Have renamed it `MessageError` — feel like it's more consistent with the `TemplateError` [thrown elsewhere](https://github.com/opengovsg/postmangovsg/blob/08da008200a75721bf0a23283f65409ad2e89d2d/backend/src/email/services/email-template.service.ts#L180-L184). 

Have added the test cases too.

File: backend/src/email/routes/tests/email-transactional.routes.test.ts
Comment (tan-zx) [reply backend/src/email/routes/tests/email-transactional.routes.test.ts:42]: As I'm adding this back into each test, I think I see your point about overgeneralisation.

Basically, for the tests that fail even before reaching the `sendEmail` stage, there is no need to mock their resolved value, so mocking resolved value for all tests is an overkill!

**Comment by tan-zx** [comment]
I have fixed the stuff mentioned in the resolved convos (have made discrete commits so it's easier to track changes).

I'm not very sure how to use the manual mock you linked and how to mock the external dependency for the malicious file test → would be great if we could find some time to go through them? 🙏 (I have previously encountered this issue and have struggled to fix it... teach me sempai)

File: backend/src/email/routes/tests/email-transactional.routes.test.ts
Comment (tan-zx) [inline backend/src/email/routes/tests/email-transactional.routes.test.ts:287]: realize that we actually insert `<br />` into the subject since we're using the same function `replaceNewLinesAndSanitize` for both subject and body. 

it doesn't affect the email sent, but just wanted to check if this is okay?

or if we should use two separate functions for sanitizing subject vs body

File: backend/src/email/routes/tests/email-transactional.routes.test.ts
Comment (tan-zx) [inline backend/src/email/routes/tests/email-transactional.routes.test.ts:251]: as I was writing this test, I realize that we're actually replacing `\n` and `\r\n` in the subject header with `<br />` as part of the sanitization process. I thought it's weird because there shouldn't be line breaks in the subject line at all. 

conversely, it seems like we're not doing this for the body? I'm not quite sure what's going on there (the comments suggest that it's doing this so that things render properly on the frontend, but I still don't fully understand), but I don't think it's worth the effort to fix.

File: backend/src/core/services/file-attachment.service.ts
Comment (stanleynguyen) [inline backend/src/core/services/file-attachment.service.ts:76]: This is still exported solely for mocking. Please fix

File: backend/src/email/middlewares/email-transactional.middleware.ts
Comment (stanleynguyen) [reply backend/src/email/middlewares/email-transactional.middleware.ts:31]: Unresolved comment

**Comment by tan-zx** [comment]
Fixed the tests per PR comments! Remaining unresolved convos are either just random comments or suggestions to be taken up in subsequent PRs.

File: backend/src/email/services/email-transactional.service.ts
Comment (tan-zx) [inline backend/src/email/services/email-transactional.service.ts:71]: I moved this from the utils folder into `EmailService` — not 100% sure whether this is correct, hence flagging this.

**Review by stanleynguyen — APPROVED** [review]

File: backend/src/email/routes/tests/email-transactional.routes.test.ts
Comment (stanleynguyen) [reply backend/src/email/routes/tests/email-transactional.routes.test.ts]: Yup we will need to handle that error, but not in this PR. Don't think it's a big issue (yet) since we're still manually onboarding our B2B users and not leaving then to self-service yet

**Pull request merged**
